### PR TITLE
Adding the MINGDA D2

### PR DESCRIPTION
**Pull request by cataclism**
Hi! Cataclism here! I've worked in order to add the MINGDA D2 to Cura. The base is currently missing but i'll add it soon =)

### CODE REVIEW

**Comment by nallath** [comment]
I'm a bit confused by the commit history of this PR. Why does it make all kinds of changes to the artillery_base?

**Comment by cataclism** [comment]
@nallath i don't know why all the history of changes I made in the past are attacched to the PR. Only the last point of the list is the one of interest, all the previous are from my work for the Artillery printers and it has already been approved

**Comment by cataclism** [comment]
I don't know how to remove it, if you can tell me how to, i'll do it @nallath

**Comment by Ghostkeeper** [comment]
We can squash the commits when merging this.

You can also squash the commits yourself if you'd like to help. There are some instructions [online](https://gitbetter.substack.com/p/how-to-squash-git-commits). The basic idea is to do `git rebase -i master` while on your branch, which comes up with a file editor where you can say which commits need to be squashed into the previous commit. Then afterwards you'd need to force-push.

**Comment by Ghostkeeper** [comment]
If the profiles for these printers are the same as the Artillery printers, I'd like to propose a change to set the following metadata on all of the new printers:

    "quality_definition": "artillery_base"

This tells Cura that you're using the Artillery's profiles. This has three consequences:

* If the Artillery profiles get updated, Mingda profiles will get updated too. This could be good (you get improvements made by the Artillery users as well) or bad (those improvements don't necessarily apply to your printer in some cases, although it might apply in many cases).
* You can then delete these quality profiles. Cura won't need to read these extra ~80 files at every launch, which will save an extra I/O block or two to read from disk.
* Material profiles that apply to Artillery printers will also apply to Mingda. This is something that is currently missing from your PR I think.

You can still override settings in the definition files and nozzle files, but you'd share the quality and material profiles with Artillery then.

**Comment by cataclism** [comment]
I'll try to squash, hope i can do it! 
I'd like to keep them separate since they are different machines. I'll work on it

**Comment by Ghostkeeper** [comment]
> I'd like to keep them separate since they are different machines.

I'd like to deduplicate them if their profiles are the same. If they need to be different that's fine, but if they might be different in the future we can always split the profiles then and deduplicate it now.

**Comment by cataclism** [comment]
Ok @Ghostkeeper , got it. I'm working with the community in order to make the transition to the profiles as smooth as possible. So  changhes are coming in the next days

**Comment by cataclism** [comment]
@Ghostkeeper  first of all happy new year!!
I've talked with the community and the profiles must be separated. 
So If there aren't any problems, you can merge.

File: resources/definitions/mingda_d2.def.json [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+        "machine_width": { "default_value": 230 },
+        "machine_depth": { "default_value": 230 },
+        "machine_height": { "default_value": 260 },
+        "gantry_height": { "value": 25 }
+
+    },
+    "metadata": {
+        "quality_definition": "mingda_base",
+        "visible": true,
+		"platform": "mingda_d2_base.stl",
```
Comment (konskarm) [inline resources/definitions/mingda_d2.def.json:16]:
```suggestion
        "platform": "mingda_d2_base.stl",
```
Use spaces only, not tabs

File: resources/quality/mingda/PETG/mingda_0.3_PETG_standard.inst.cfg [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+setting_version = 16
+type = quality
+quality_type = standard
+material = generic_petg
+variant = 0.3mm Nozzle
+
+[values]
+speed_layer_0 = 15
+wall_thickness = =line_width*4
+#retraction_extra_prime_amount = 0.5
```
Comment (konskarm) [inline resources/quality/mingda/PETG/mingda_0.3_PETG_standard.inst.cfg:16]: Remove comment

```suggestion
```

File: resources/quality/mingda/PETG/mingda_0.3_PETG_super.inst.cfg [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+setting_version = 16
+type = quality
+quality_type = super
+material = generic_petg
+variant = 0.3mm Nozzle
+
+[values]
+speed_layer_0 = 15
+wall_thickness = =line_width*4
+#retraction_extra_prime_amount = 0.5
```
Comment (konskarm) [inline resources/quality/mingda/PETG/mingda_0.3_PETG_super.inst.cfg:16]: Remove comment

```suggestion
```

File: resources/quality/mingda/PETG/mingda_0.4_PETG_adaptive.inst.cfg [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+setting_version = 16
+type = quality
+quality_type = adaptive
+material = generic_petg
+variant = 0.4mm Nozzle
+
+[values]
+speed_layer_0 = 15
+wall_thickness = =line_width*4
+#retraction_extra_prime_amount = 0.5
```
Comment (konskarm) [inline resources/quality/mingda/PETG/mingda_0.4_PETG_adaptive.inst.cfg:16]: Remove comment

```suggestion
```

File: resources/quality/mingda/PETG/mingda_0.4_PETG_low.inst.cfg [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+setting_version = 16
+type = quality
+quality_type = low
+material = generic_petg
+variant = 0.4mm Nozzle
+
+[values]
+speed_layer_0 = 15
+wall_thickness = =line_width*4
+#retraction_extra_prime_amount = 0.5
```
Comment (konskarm) [inline resources/quality/mingda/PETG/mingda_0.4_PETG_low.inst.cfg:16]: Remove comment

```suggestion
```

File: resources/quality/mingda/PETG/mingda_0.4_PETG_standard.inst.cfg [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+setting_version = 16
+type = quality
+quality_type = standard
+material = generic_petg
+variant = 0.4mm Nozzle
+
+[values]
+speed_layer_0 = 15
+wall_thickness = =line_width*4
+#retraction_extra_prime_amount = 0.5
```
Comment (konskarm) [inline resources/quality/mingda/PETG/mingda_0.4_PETG_standard.inst.cfg:16]: Remove comment

```suggestion
```

File: resources/quality/mingda/PETG/mingda_0.4_PETG_super.inst.cfg [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+setting_version = 16
+type = quality
+quality_type = super
+material = generic_petg
+variant = 0.4mm Nozzle
+
+[values]
+speed_layer_0 = 15
+wall_thickness = =line_width*4
+#retraction_extra_prime_amount = 0.5
```
Comment (konskarm) [inline resources/quality/mingda/PETG/mingda_0.4_PETG_super.inst.cfg:16]: Remove comment

```suggestion
```

File: resources/quality/mingda/PETG/mingda_0.5_PETG_adaptive.inst.cfg [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+setting_version = 16
+type = quality
+quality_type = adaptive
+material = generic_petg
+variant = 0.5mm Nozzle
+
+[values]
+speed_layer_0 = 15
+wall_thickness = =line_width*4
+#retraction_extra_prime_amount = 0.5
```
Comment (konskarm) [inline resources/quality/mingda/PETG/mingda_0.5_PETG_adaptive.inst.cfg:16]: Remove comment

```suggestion
```

File: resources/quality/mingda/PETG/mingda_0.5_PETG_low.inst.cfg [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+setting_version = 16
+type = quality
+quality_type = low
+material = generic_petg
+variant = 0.5mm Nozzle
+
+[values]
+speed_layer_0 = 15
+wall_thickness = =line_width*4
+#retraction_extra_prime_amount = 0.5
```
Comment (konskarm) [inline resources/quality/mingda/PETG/mingda_0.5_PETG_low.inst.cfg:16]: Remove comment

```suggestion
```

File: resources/quality/mingda/PETG/mingda_0.5_PETG_standard.inst.cfg [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+setting_version = 16
+type = quality
+quality_type = standard
+material = generic_petg
+variant = 0.5mm Nozzle
+
+[values]
+speed_layer_0 = 15
+wall_thickness = =line_width*4
+#retraction_extra_prime_amount = 0.5
```
Comment (konskarm) [inline resources/quality/mingda/PETG/mingda_0.5_PETG_standard.inst.cfg:16]: Remove comment

```suggestion
```

File: resources/quality/mingda/PETG/mingda_0.5_PETG_super.inst.cfg [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+setting_version = 16
+type = quality
+quality_type = super
+material = generic_petg
+variant = 0.5mm Nozzle
+
+[values]
+speed_layer_0 = 15
+wall_thickness = =line_width*4
+#retraction_extra_prime_amount = 0.5
```
Comment (konskarm) [inline resources/quality/mingda/PETG/mingda_0.5_PETG_super.inst.cfg:16]: Remove comment

```suggestion
```

File: resources/quality/mingda/PETG/mingda_0.6_PETG_standard.inst.cfg [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+setting_version = 16
+type = quality
+quality_type = standard
+material = generic_petg
+variant = 0.6mm Nozzle
+
+[values]
+speed_layer_0 = 15
+wall_thickness = =line_width*3
+#retraction_extra_prime_amount = 0.5
```
Comment (konskarm) [inline resources/quality/mingda/PETG/mingda_0.6_PETG_standard.inst.cfg:16]: Remove comment

```suggestion
```

File: resources/quality/mingda/PETG/mingda_0.8_PETG_draft.inst.cfg [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+setting_version = 16
+type = quality
+quality_type = draft
+material = generic_petg
+variant = 0.8mm Nozzle
+
+[values]
+speed_layer_0 = 15
+wall_thickness = =line_width*3
+#retraction_extra_prime_amount = 0.5
```
Comment (konskarm) [inline resources/quality/mingda/PETG/mingda_0.8_PETG_draft.inst.cfg:16]: Remove comment

```suggestion
```

File: resources/quality/mingda/PETG/mingda_1.0_PETG_draft.inst.cfg [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+setting_version = 16
+type = quality
+quality_type = draft
+material = generic_petg
+variant = 1.0mm Nozzle
+
+[values]
+speed_layer_0 = 15
+wall_thickness = =line_width*3
+#retraction_extra_prime_amount = 0.5
```
Comment (konskarm) [inline resources/quality/mingda/PETG/mingda_1.0_PETG_draft.inst.cfg:16]: Remove comment

```suggestion
```

File: resources/definitions/mingda_base.def.json [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+        "retraction_retract_speed": {
+            "maximum_value_warning": "machine_max_feedrate_e if retraction_enable else float('inf')",
+            "maximum_value": 200
+        },
+        "retraction_prime_speed": {
+            "maximum_value_warning": "machine_max_feedrate_e if retraction_enable else float('inf')",
+            "maximum_value": 200
+        },
+        
+        "retraction_hop_enabled": { "value": "true" },
```
Comment (konskarm) [inline resources/definitions/mingda_base.def.json:212]:
```suggestion
        "retraction_hop_enabled": { "value": true },
```
`true` or `"True"` works, but not `"true"`.

File: resources/definitions/mingda_base.def.json [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+        "speed_prime_tower": { "value": "speed_topbottom" },
+        "speed_support": { "value": "speed_wall_0" },
+        "speed_support_interface": { "value": "speed_topbottom" },
+        "speed_z_hop": { "value": 5 },
+
+        "skirt_brim_speed": { "value": "speed_layer_0" },
+
+        "line_width": { "value": "machine_nozzle_size * 1.1" },
+
+        "optimize_wall_printing_order": { "value": "True" },
```
Comment (konskarm) [inline resources/definitions/mingda_base.def.json:177]:
```suggestion
        "optimize_wall_printing_order": { "value": true },
```
`"True"` will work, but it's better to use `true` instead.

File: resources/definitions/mingda_base.def.json [hunk 17]
```diff
@@ -0,0 +1,265 @@
...
+
+        "material_initial_print_temperature": { "value": "material_print_temperature" },
+        "material_final_print_temperature": { "value": "material_print_temperature" },
+        "material_flow": { "value": 100 },
+        "travel_compensate_overlapping_walls_0_enabled": { "value": "False" },
```
Comment (konskarm) [inline resources/definitions/mingda_base.def.json:182]:
```suggestion
        "travel_compensate_overlapping_walls_0_enabled": { "value": false },
```
`"False"` will work, but it's better to use `false` instead.

**Review by konskarm — CHANGES_REQUESTED** [review]

**Review by cataclism — COMMENTED** [review]
Hope I did everything right! If there are any problems tell me @konskarm

**Comment by konskarm** [comment]
Alright, since it is decided that the MINGDA definition should be completely separate from the Artillery ones, I will accept this PR. Just make sure to re-use the mingda_base, or else we are adding 88 files for a single printer, which is a bit excessive.

**Pull request merged**